saleor/graphql/product/mutations/products.py

```diff
@@ -469,6 +476,189 @@ class ProductCreateInput(ProductInput):
+
+        - ensure all required attributes are passed
+        - ensure the values are correct for a product
+        """
+        supplied_attribute_pk = []
+        for attribute, values in cleaned_input:
+            validate_attribute_input_for_product(attribute, values)
+            supplied_attribute_pk.append(attribute.pk)
+
+        # Asserts all required attributes are supplied
+        missing_required_filter = Q(value_required=True) & ~Q(
+            pk__in=supplied_attribute_pk
+        )
+
+        if qs.filter(missing_required_filter).exists():
+            raise ValidationError(
+                "All attributes flagged as having a value required must be supplied."
+            )
+
+    @classmethod
+    def _check_input_for_variant(cls, cleaned_input: T_INPUT_MAP, qs: QuerySet):
+        """Check the cleaned attribute input for a variant.
+
+        An Attribute queryset is supplied.
+
+        - ensure all attributes are passed
+        - ensure the values are correct for a variant
+        """
+        if len(cleaned_input) != qs.count():
+            raise ValidationError("All attributes must take a value")
+
+        for attribute, values in cleaned_input:
+            validate_attribute_input_for_variant(attribute, values)
+
+    @classmethod
+    def clean_input(cls, raw_input: dict, attributes_qs: QuerySet) -> T_INPUT_MAP:
+        """Resolve and prepare the input for further checks.
+
+        :param raw_input: The user's attributes input.
+        :param attributes_qs: A queryset of attributes, must prefetch attribute values.
+
+        :raises ValidationError: contain the message.
+        :return: The resolved data
+        """
+
+        # Mapping to associate the input values back to the resolved attribute nodes
+        pks = {}
+        slugs = {}
+
+        # Temporary storage of the passed ID for error reporting
+        global_ids = []
+
+        for attribute_input in raw_input:
+            global_id = attribute_input.get("id")
+            slug = attribute_input.get("slug")
+            values = attribute_input["values"]
+
+            if global_id:
+                internal_id = cls._resolve_attribute_global_id(global_id)
+                global_ids.append(global_id)
+                pks[internal_id] = values
+            elif slug:
+                slugs[slug] = values
+            else:
+                raise ValidationError("You must whether supply an ID or a slug")
+
+        attributes = cls._get_attribute_nodes(
+            attributes_qs, global_ids=global_ids, pks=pks.keys(), slugs=slugs.keys()
+        )
+        cleaned_input = []
+        for attribute in attributes:
+            key = pks.get(attribute.pk, None)
+
+            if key is None:
+                key = slugs[attribute.slug]
+
+            cleaned_input.append((attribute, key))
+        return cleaned_input
+
+    @classmethod
+    def validate_input(cls, cleaned_input: T_INPUT_MAP, attribute_qs, is_variant: bool):
```

Comment:
Not sure how useful this method is - since you have to pass a right flag to trigger to flow you want, you could also just call the right method that is called underneath.